python/paddle/distributed/auto_parallel/static/tuner/pir_rule_based_tuner.py

```diff
@@ -0,0 +1,1487 @@
+        attn_output = attn_output.transpose([0, 2, 1, 3])
+
+        attn_output = attn_output.reshape([bsz, q_len, head_dim * num_heads])
+
+        return attn_output
+
+
+# Llama
+@register_pir_pattern
+class PIRMLPPattern(PIRBasePattern):
+    """MLP pattern"""
+
+    name = "mlp"
+
+    def __init__(self):
+        super().__init__()
+
+    def build(self):
+        # # # build program # # #
+        paddle.enable_static()
+        # program init
+        start_program, main_program = (
+            paddle.static.Program(),
+            paddle.static.Program(),
+        )
+        # data init
+        batch_size = 4
+        seq_length = 1024
+        hidden_size = 2048
+        intermediate_size = 4096
+        hidden_states_shape = [batch_size, seq_length, hidden_size]
+        hidden_states = paddle.randn(hidden_states_shape)
+        up_weight_shape = [hidden_size, intermediate_size]
+        down_weight_shape = [intermediate_size, hidden_size]
+
+        # program construction
+        with paddle.static.program_guard(main_program, start_program):
+            hidden_states = paddle.static.data(
+                'hidden_states', hidden_states_shape, hidden_states.dtype
+            )
+            gate_weight = paddle.create_parameter(up_weight_shape, "float32")
+            up_weight = paddle.create_parameter(up_weight_shape, "float32")
+            down_weight = paddle.create_parameter(down_weight_shape, "float32")
+
+            # qkv (not fused)
+            gate = paddle.matmul(hidden_states, gate_weight)
+            up = paddle.matmul(hidden_states, up_weight)
+            tmp = paddle.incubate.nn.functional.swiglu(gate, up)
+            out = paddle.matmul(tmp, down_weight)
+
+        self.pir_program = main_program
+        # print(f"in MLPPattern, program is {self.pir_program}")
+        paddle.disable_static()
+
+        up_linear_dist_infos = MpDistInfos("column")
+        down_linear_dist_infos = MpDistInfos("row")
+        # up_linear_dist_infos = MpDistInfos()
+        # down_linear_dist_infos = MpDistInfos()
+
+        # # # build ops dist infos # # #
+        ops_dist_infos = {
+            (4,): up_linear_dist_infos,
+            (5,): up_linear_dist_infos,
+            (7,): down_linear_dist_infos,
+        }
+        self.ops_dist_infos = ops_dist_infos
+
+
+# Llama
+@register_pir_pattern
+class PIRDecoderLayerPattern(PIRBasePattern):
+    """Decoder layer pattern"""
+
+    name = "decoder_layer"
+
+    def __init__(self):
+        super().__init__()
+
+    def build(self):
+        # # # build program # # #
+        paddle.enable_static()
+        # program init
+        start_program, main_program = (
+            paddle.static.Program(),
+            paddle.static.Program(),
+        )
+        # data init
+        batch_size = 4
+        seq_length = 1024
+        num_heads = 32
+        head_size = 64
+        hidden_size = num_heads * head_size
+        intermediate_size = 4096
```

Comment:
it is better to move all these configurations which required for each subclass to the parent class?

Reply:
shall we set the shape smaller without losing generality and reducing memory usage.

Reply:
done